### ISSUE TITLE
Tighten up string type representations to prevent illegal values

### ISSUE THREAD

**Issue by cpu** [rustls/rcgen#181]
In several places Rcgen defines representations of ASN.1 string types, but in a way that doesn't enforce the values meet the restrictions imposed by the type in use. 

For example, consider the `DnValue` enum with `TeletexString`, `PrintableString`, `UniversalString`, `Utf8String` and `BmpString` variants. Those variants express their values as either `String` or `Vec<u8>`, but in many cases the ASN.1 definitions for these string types introduce further restrictions. As one example, a `PrintableString` can only contain `A-Z, a-z, 0-9, '()+,-./:=?` and `<SPACE>` where as a Rust `String` can contain any valid UTF-8.

Because of this mismatch invalid values can be expressed in `CertificateParams`, and won't be caught by any validation. In the cases where we use a yasna writer and `write_bytes` generically, we'll emit an invalid encoding for the type in use. In the cases where we use a more specific yasna helper like `write_ia5_string` or `write_printable_string`, invalid values will cause a panic.

I think we should look at creating a different representation that can impose more validation in order to reject invalid values at construction time as opposed to emitting invalid encodings or panicing when serializing.

**Comment by Tudyx** [comment]
Does using newtype would make the job?

Something in the idea of the [nutype crate](https://github.com/greyblake/nutype) but without using derive macro

**Comment by djc** [comment]
Yeah, we could probably do newtypes for each of the inner variants, and have `TryFrom` impls for them?

**Comment by cpu** [comment]
That sounds like the right approach to me :+1:

**Comment by Tudyx** [comment]
The created types will look like a lot to the one from the [der::asn1 module](https://docs.rs/der/latest/der/asn1/index.html) of the `der` crate. See for instance the [`PrintableString struct`](https://docs.rs/der/latest/der/asn1/struct.PrintableString.html).
Should we consider adding the `der` crate as a dependency or it's not worth it?

**Comment by est31** [comment]
I don't think we should depend on the der crate for that alone. It has a huge list of dependencies of its own.

**Comment by Tudyx** [comment]
Actually, it has no dependency for our use case. All the dependencies listed in [their ` Cargo.toml`](https://github.com/RustCrypto/formats/blob/master/der/Cargo.toml) are optional, and we only need the `alloc` feature that doesn't activate any. I double-check with `cargo tree` to be sure.
